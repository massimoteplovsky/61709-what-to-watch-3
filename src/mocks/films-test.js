export const films = [
  {
    id: 1,
    title: `Terminator 2. Judgment day`,
    genre: [`Action`],
    year: 1991,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 2,
    title: `1+1`,
    genre: [`Drama`, `Comedy`],
    year: 2012,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 3,
    title: `Joker`,
    genre: [`Drama`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 4,
    title: `Once upon a time in Hollywood`,
    genre: [`Drama`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 5,
    title: `Ford vs Ferrari`,
    genre: [`Drama`, `History`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 6,
    title: `1917`,
    genre: [`Drama`, `History`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 7,
    title: `Parasite`,
    genre: [`Drama`, `Comedy`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  },
  {
    id: 8,
    title: `Irishman`,
    genre: [`drama`, `criminal`],
    year: 2019,
    promoPoster: ``,
    poster: ``,
    cover: ``,
    description: ``,
    rating: 8.5,
    votes: 12000,
    director: ``,
    starring: []
  }
];

export const film = {
  id: 1,
  title: `Terminator 2. Judgment day`,
  genre: [`Action`],
  year: 1991,
  promoPoster: ``,
  poster: ``,
  cover: ``,
  description: ``,
  rating: 8.5,
  votes: 12000,
  director: ``,
  starring: []
};

export const promoFilmInfo = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  year: 2014
};
