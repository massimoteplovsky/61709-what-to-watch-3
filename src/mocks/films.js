export default [
  {
    id: 1,
    title: `Terminator 2. Judgment day`,
    poster: `http://i4.ytimg.com/vi/HgV7-MJwUBw/mqdefault.jpg`
  },
  {
    id: 2,
    title: `1+1`,
    poster: `https://i.pinimg.com/736x/76/44/c0/7644c07844dd4cb28d2c289bfa6e0dec--the-intouchables-comedy-movies.jpg`
  },
  {
    id: 3,
    title: `Joker`,
    poster: `https://all.culture.ru/uploads/9b743d135092e110a8e55a3dc3ad5bc4_w375_h250.jpeg`
  },
  {
    id: 4,
    title: `Once upon a time in Hollywood`,
    poster: `https://avatars.mds.yandex.net/get-pdb/1945878/3281b62e-4539-4eb7-a26b-fa68f6572bea/s375`
  },
  {
    id: 5,
    title: `Ford vs Ferrari`,
    poster: `https://sun9-28.userapi.com/c206816/v206816832/1771c/3p_wj8sfMz8.jpg`
  },
  {
    id: 6,
    title: `1917`,
    poster: `https://sun3-13.userapi.com/c857724/v857724955/14e882/u5CFMvbvX3I.jpg`
  },
  {
    id: 7,
    title: `Parasite`,
    poster: `http://m.kino-teatr.ru/art/5395/pv_75950.jpg`
  },
  {
    id: 8,
    title: `Irishman`,
    poster: `https://sun9-31.userapi.com/c850608/v850608676/40eb4/ojB5App6exk.jpg`
  },
];
