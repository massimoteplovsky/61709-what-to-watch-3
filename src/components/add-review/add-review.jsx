import React, {Fragment} from "react";
import {connect} from "react-redux";
import {PropValidator} from "../../prop-validator/prop-validator.js";
import {PropTypes} from "prop-types";
import {Link, Redirect} from "react-router-dom";
import history from "../../history.js";
import {getFilm} from "../../selectors/films/films.js";
import {getUserInfo} from "../../selectors/user/user.js";
import {sendReview} from "../../actions/action-creators/film-actions/film-actions.js";
import {BASE_URL} from "../../consts.js";
import withReviewForm from "../../hocs/with-review-form/with-review-form.js";

const ratings = new Array(5).fill(``).map((_, i) => i + 1);

const AddReview = ({
  filmInfo,
  userInfo,
  comment,
  rating,
  isDisabled,
  isCommentValid,
  isRatingValid,
  isFormValid,
  onChange,
  onDisable,
  onSendReview,
  onSendFormSuccess,
  formMessage
}) => {

  if (!filmInfo) {
    return <Redirect to="/"/>;
  }

  const {
    id,
    backgroundImage,
    posterImage,
    name
  } = filmInfo;

  const {
    avatarUrl,
    name: userName
  } = userInfo;

  const onSubmitHandler = (event) => {
    event.preventDefault();

    if (!isFormValid) {
      return;
    }

    onDisable(true);

    onSendReview({rating, comment}, id, onSendFormSuccess);
  };

  return (
    <section className="movie-card movie-card--full">
      <div className="movie-card__header">
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={name} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header">
          <div className="logo">
            <Link to="/" className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <nav className="breadcrumbs">
            <ul className="breadcrumbs__list">
              <li className="breadcrumbs__item">
                <Link to={`/films/${id}`} className="breadcrumbs__link">{name}</Link>
              </li>
              <li className="breadcrumbs__item">
                <Link to={`/films/${id}/review`} className="breadcrumbs__link">Add review</Link>
              </li>
            </ul>
          </nav>

          <div className="user-block">
            <div className="user-block__avatar">
              <img
                src={`${BASE_URL}${avatarUrl}`}
                alt={userName} width="63" height="63"
                onClick={() => history.push(`/myList`)}
              />
            </div>
          </div>
        </header>

        <div className="movie-card__poster movie-card__poster--small">
          <img src={posterImage} alt={name} width="218" height="327" />
        </div>
      </div>

      <div className="add-review">
        <form
          onSubmit={onSubmitHandler}
          action="#"
          className="add-review__form"
        >
          <div className="rating">
            <div className="rating__stars">
              {
                ratings.map((it) => {
                  return (
                    <Fragment key={it}>
                      <input
                        className="rating__input"
                        id={`star-${it}`}
                        type="radio"
                        name="rating"
                        value={it}
                        checked={it === rating}
                        onChange={onChange}
                        disabled={isDisabled}
                      />
                      <label
                        className="rating__label"
                        htmlFor={`star-${it}`}
                        style={{opacity: (it <= rating) ? 1 : 0.5}}
                      >
                        Rating {it}
                      </label>
                    </Fragment>
                  );
                })
              }
            </div>
          </div>

          <div className="add-review__text">
            <textarea
              className="add-review__textarea"
              name="review-text"
              id="review-text"
              placeholder="Review text"
              value={comment}
              onChange={onChange}
              disabled={isDisabled}
              style={{opacity: isDisabled ? 0.3 : 1}}
            >
            </textarea>
            <div className="add-review__submit">
              <button
                className="add-review__btn"
                type="submit"
                disabled={!isRatingValid || !isCommentValid || isDisabled ? true : false}
                style={{opacity: !isRatingValid || !isCommentValid || isDisabled ? 0.3 : 1}}
              >
                Post
              </button>
            </div>
          </div>
        </form>
        {
          formMessage
          &&
          <p style={{
            fontSize: `30px`,
            textAlign: `center`
          }}>{formMessage}</p>
        }
      </div>

    </section>
  );
};

AddReview.propTypes = {
  filmInfo: PropValidator.FILM_INFO,
  userInfo: PropValidator.USER_INFO,
  comment: PropTypes.string.isRequired,
  rating: PropTypes.number.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  isCommentValid: PropTypes.bool.isRequired,
  isRatingValid: PropTypes.bool.isRequired,
  isFormValid: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  onDisable: PropTypes.func.isRequired,
  onSendReview: PropTypes.func.isRequired,
  onSendFormSuccess: PropTypes.func.isRequired,
  formMessage: PropTypes.string.isRequired
};

const mapStateToProps = (state, ownProps) => {
  return {
    filmInfo: getFilm(state, ownProps.filmID),
    userInfo: getUserInfo(state)
  };
};

const mapDispatchToProps = (dispatch) => ({
  onSendReview(reviewInfo, id, onSendFormSuccess) {
    dispatch(sendReview(reviewInfo, id, onSendFormSuccess));
  }
});

export {AddReview};
export default connect(mapStateToProps, mapDispatchToProps)(withReviewForm(AddReview));
