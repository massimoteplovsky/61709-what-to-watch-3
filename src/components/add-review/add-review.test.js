import React from "react";
import renderer from "react-test-renderer";
import {AddReview} from "./add-review.jsx";
import {Router} from "react-router-dom";
import {film, userInfo} from "../../mocks/films-test";
import history from "../../history.js";

it(`<AddReview/> component renders correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <AddReview
            filmInfo={film}
            userInfo={userInfo}
            comment={``}
            rating={4}
            isDisabled={false}
            isCommentValid={true}
            isRatingValid={true}
            isFormValid={true}
            onChange={() => {}}
            onDisable={() => {}}
            onSendReview={() => {}}
            onSendFormSuccess={() => {}}
            formMessage={``}
          />
        </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
