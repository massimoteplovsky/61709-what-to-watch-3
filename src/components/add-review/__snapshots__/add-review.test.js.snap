// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddReview/> component renders correctly 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Film 1"
        src=""
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1"
              onClick={[Function]}
            >
              Film 1
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <img
            alt=""
            height="63"
            onClick={[Function]}
            src="https://htmlacademy-react-2.appspot.com"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="Film 1"
        height="327"
        src=""
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    <form
      action="#"
      className="add-review__form"
      onSubmit={[Function]}
    >
      <div
        className="rating"
      >
        <div
          className="rating__stars"
        >
          <input
            checked={false}
            className="rating__input"
            disabled={false}
            id="star-1"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={1}
          />
          <label
            className="rating__label"
            htmlFor="star-1"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            Rating 
            1
          </label>
          <input
            checked={false}
            className="rating__input"
            disabled={false}
            id="star-2"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={2}
          />
          <label
            className="rating__label"
            htmlFor="star-2"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            Rating 
            2
          </label>
          <input
            checked={false}
            className="rating__input"
            disabled={false}
            id="star-3"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={3}
          />
          <label
            className="rating__label"
            htmlFor="star-3"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            Rating 
            3
          </label>
          <input
            checked={true}
            className="rating__input"
            disabled={false}
            id="star-4"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={4}
          />
          <label
            className="rating__label"
            htmlFor="star-4"
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            Rating 
            4
          </label>
          <input
            checked={false}
            className="rating__input"
            disabled={false}
            id="star-5"
            name="rating"
            onChange={[Function]}
            type="radio"
            value={5}
          />
          <label
            className="rating__label"
            htmlFor="star-5"
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            Rating 
            5
          </label>
        </div>
      </div>
      <div
        className="add-review__text"
      >
        <textarea
          className="add-review__textarea"
          disabled={false}
          id="review-text"
          name="review-text"
          onChange={[Function]}
          placeholder="Review text"
          style={
            Object {
              "opacity": 1,
            }
          }
          value=""
        />
        <div
          className="add-review__submit"
        >
          <button
            className="add-review__btn"
            disabled={false}
            style={
              Object {
                "opacity": 1,
              }
            }
            type="submit"
          >
            Post
          </button>
        </div>
      </div>
    </form>
    
  </div>
</section>
`;
